GitCommands/Git/GitGpgController.cs

```diff
@@ -0,0 +1,219 @@
+
+        /// <summary>
+        /// Obtain the number of tag on current git revision.
+        /// </summary>
+        /// <returns>Returns the number of annotated or signed tag on current revision.</returns>
+        public int NumberOfTag { get; private set; }
+
+        /// <summary>
+        /// Obtain the tag verification message for all the tag in current git revision 
+        /// </summary>
+        /// <returns>Full concatenated string coming from GPG analysis on all tags on current git revision.</returns>
+        public string TagVerifyMessage { get; private set; }
+
+        public GitGpgController(IGitModule module, GitRevision revision)
+        {
+            _module = module;
+            _revision = revision;
+
+            NumberOfTag = _revision.Refs.Count(x => x.IsTag && x.IsDereference);
+            TagVerifyMessage = "";
+        }
+
+        /// <summary>
+        /// Obtain the commit signature status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate the gpg status for current git revision.</returns>
+        public CommitStatus CheckCommitSign()
+        {
+            CommitStatus _cmtStatus = CommitStatus.NoSignature;
+            string _gpg = _module.RunGitCmd($"log --pretty=\"format:%G?\" -1 {_revision.Guid}");
+
+            switch (_gpg)
+            {
+                case GOOD_SIGN:         // "G" for a good (valid) signature
+                    _cmtStatus = CommitStatus.GoodSignature;
+                    break;
+                case BAD_SIGN:          // "B" for a bad signature
+                case UNK_SIGN_VALIDITY: // "U" for a good signature with unknown validity 
+                case EXPIRED_SIGN:      // "X" for a good signature that has expired
+                case EXPIRED_SIGN_KEY:  // "Y" for a good signature made by an expired key
+                case REVOKED_KEY:       // "R" for a good signature made by a revoked key
+                    _cmtStatus = CommitStatus.SignatureError;
+                    break;
+                case MISSING_PUB_KEY:   // "E" if the signature cannot be checked (e.g.missing key)
+                    _cmtStatus = CommitStatus.MissingPublicKey;
+                    break;
+                case NO_SIGN:           // "N" for no signature
+                default:
+                    _cmtStatus = CommitStatus.NoSignature;
+                    break;
+            }
+
+            return _cmtStatus;
+        }
+
+
+        /// <summary>
+        /// Obtain the tag status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate if current git revision has one tag with good signature, one tag with bad signature or more than one tag.</returns>
+        public TagStatus CheckTagSign()
+        {
+            TagStatus _tagStatus = TagStatus.OneBad;
+
+            IEnumerable<IGitRef> _usefulRef = _revision.Refs.Where(x => x.IsTag && x.IsDereference);
```

Comment:
the collection needs to be enumerated (i.e. `ToList()`), otherwise it gets enumerated multiple times (line 154 and 159 or 168). and once enumerated line 154 can be changed to `if (usefulRef.Count > 1)`

Reply:
look at newest commit